wwwverifier/src/main/java/com/google/sps/servlets/RequestServlet.java

```
@@ -0,0 +1,514 @@
+    * @return Generated mdoc:// URI
+    */
+    private static String createMdocUri(Key key) {
+        KeyPair keyPair = generateKeyPair();
+        byte[] re = generateReaderEngagement(keyPair.getPublic(), key);
+  
+        // put ReaderEngagement and generated ephemeral keys into Datastore
+        setDatastoreProp(ServletConsts.RE_PROP, re, key);
+        setDatastoreProp(ServletConsts.PUBKEY_PROP, keyPair.getPublic().getEncoded(), key);
+        setDatastoreProp(ServletConsts.PRIVKEY_PROP, keyPair.getPrivate().getEncoded(), key);
+
+        String reStr = Base64.getUrlEncoder().withoutPadding().encodeToString(re).replace("\n","");
+  
+        return ServletConsts.MDOC_PREFIX + reStr;
+    }
+
+    /**
+     * Handles two distinct POST requests:
+     * (1) The first POST request sent to the servlet, which contains a MessageData message
+     * (2) The second POST request sent to the servlet, which contains a DeviceResponse message
+     * 
+     * @param request encoded CBOR message
+     * @return (1) response, containing a DeviceRequest message as an encoded byte array; 
+     * (2) response, containing a termination message with status code 20
+     */
+    @Override
+    public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+        String[] pathArr = parsePathInfo(request);
+        Key key = com.google.appengine.api.datastore.KeyFactory.stringToKey(pathArr[0]);
+        if (getNumPostRequests(key) == 0) {
+            byte[] sessionData = createDeviceRequest(getBytesFromRequest(request), key);
+            setNumPostRequests(1, key);
+            response.getOutputStream().write(sessionData);
+        } else if (getNumPostRequests(key) == 1) {
+            byte[] terminationMessage = parseDeviceResponse(getBytesFromRequest(request), key);
+            setNumPostRequests(2, key);
+            response.getOutputStream().write(terminationMessage);
+        }
```

Should we also return an error in case getNumPostRequests() isn't 0 or 1?